File: app/src/main/java/com/team2052/frckrawler/database/tables/PitData.java
package com.team2052.frckrawler.database.tables;

import com.team2052.frckrawler.database.DBManager;
import com.team2052.frckrawler.db.Metric;
import com.team2052.frckrawler.db.PitDatum;
import com.team2052.frckrawler.db.PitDatumDao;

import org.greenrobot.greendao.query.QueryBuilder;

import java.util.Date;

public class PitData extends Table<PitDatum, PitDatumDao> {
    public PitData(PitDatumDao dao, DBManager dbManager) {
        super(dao, dbManager);
    }

    public boolean insertWithSaved(PitDatum pitDatum) {
        pitDatum.setId(null);
        QueryBuilder<PitDatum> pitDataQueryBuilder = dao.queryBuilder();
        pitDataQueryBuilder.where(PitDatumDao.Properties.Robot_id.eq(pitDatum.getRobot_id()));
        pitDataQueryBuilder.where(PitDatumDao.Properties.Metric_id.eq(pitDatum.getMetric_id()));
        long count = pitDataQueryBuilder.count();

        if (count > 0) {
            PitDatum unique = pitDataQueryBuilder.unique();
            if (unique.getLast_updated().getTime() <= System.currentTimeMillis()) {
                unique.setLast_updated(new Date());
                unique.setData(pitDatum.getData());
                dao.update(unique);
            }
            return false;
        } else {
            pitDatum.setLast_updated(new Date());
            dao.insert(pitDatum);
            return true;
        }
    }

    public QueryBuilder<PitDatum> query(Long robot_id, Long metric_id, Long event_id) {
        QueryBuilder<PitDatum> queryBuilder = getQueryBuilder();
        if (robot_id != null)
            queryBuilder.where(PitDatumDao.Properties.Robot_id.eq(robot_id));
        if (metric_id != null)
            queryBuilder.where(PitDatumDao.Properties.Metric_id.eq(metric_id));
        if (event_id != null)
            queryBuilder.where(PitDatumDao.Properties.Event_id.eq(event_id));
        return queryBuilder;
    }

    public Metric getMetric(PitDatum pitDatum) {
        return dbManager.getMetricsTable().load(pitDatum.getMetric_id());
    }

    @Override
    public PitDatum load(long id) {
        return dao.load(id);
    }

    @Override
    public void delete(PitDatum model) {
        dao.delete(model);
    }

    @Override
    public void insert(PitDatum model) {
        insertWithSaved(model);
    }
}

File: app/src/main/java/com/team2052/frckrawler/data/tables/PitData.kt
package com.team2052.frckrawler.data.tables

import com.team2052.frckrawler.data.DBManager
import com.team2052.frckrawler.models.PitDatum
import com.team2052.frckrawler.models.PitDatumDao
import org.greenrobot.greendao.query.QueryBuilder
import java.util.*

class PitData(dao: PitDatumDao, dbManager: DBManager) : AbstractTable<PitDatum, PitDatumDao>(dao, dbManager) {
    fun query(team_id: Long?, metric_id: Long?): QueryBuilder<PitDatum> {
        val queryBuilder = queryBuilder
        if (team_id != null)
            queryBuilder.where(PitDatumDao.Properties.Team_id.eq(team_id))
        if (metric_id != null)
            queryBuilder.where(PitDatumDao.Properties.Metric_id.eq(metric_id))
        return queryBuilder
    }

    fun insertWithSaved(pitDatum: PitDatum): Boolean {
        pitDatum.id = null
        val pitDataQueryBuilder = query(pitDatum.team_id, pitDatum.metric_id)
        val count = pitDataQueryBuilder.count()
        if (count > 0) {
            val unique = pitDataQueryBuilder.unique()
            if (unique.last_updated.time <= System.currentTimeMillis()) {
                unique.last_updated = Date()
                unique.data = pitDatum.data
                dao.update(unique)
            }
            return false
        } else {
            pitDatum.last_updated = Date()
            dao.insert(pitDatum)
            return true
        }
    }
}